{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OmicIDX analytical capabilities\n",
    "\n",
    "The OmicIDX REST API has capabilities to perform\n",
    "very fast summaries of query results. In practice, this functionality\n",
    "is useful for understanding large sets of OmicIDX without\n",
    "having to return all query results. In this notebook, I will give examples\n",
    "of how to use this functionality to perform basic reporting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "\n",
    "class OResult(object):\n",
    "    def __init__(self, results):\n",
    "        self._data = results\n",
    "     \n",
    "    @property\n",
    "    def hits(self) -> dict:\n",
    "        \"\"\"\"\"\"\n",
    "        return self._data['hits']\n",
    "    \n",
    "    @property\n",
    "    def total(self):\n",
    "        \"\"\"Returns the total number of records found\"\"\"\n",
    "        return self._data['stats']['total']\n",
    "    \n",
    "    @property\n",
    "    def facets(self):\n",
    "        return self._data['facets']\n",
    "    \n",
    "    @property\n",
    "    def facet_names(self):\n",
    "        return list(self._data['facets'].keys())\n",
    "    \n",
    "    def cursor(self):\n",
    "        return self._data['cursor']\n",
    "    \n",
    "\n",
    "class OClient(object):\n",
    "    def __init__(self, base_url=\"https://api.omicidx.cancerdatasci.org\"):\n",
    "        self.base_url = base_url\n",
    "        \n",
    "    def get(self, path, **query_parameters):\n",
    "        try:\n",
    "            return OResult(results = requests.get(self.base_url+path, params = query_parameters).json())\n",
    "        except:\n",
    "            print(\"error\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total hits:  226283\n",
      "Titles.............\n",
      "Bacillus subtilis subsp. natto BEST195 genome sequencing project\n",
      "Model organism for prokaryotic cell differentiation and development\n",
      "Comprehensive identification and characterization of the nucleosome             structure\n",
      "Comprehensive identification and characterization of the transcripts, their             expression levels and sub-cellular localizations\n",
      "Comprehensive identification and characterization of the transcripts, their             expression levels and sub-cellular localizations\n",
      "Comprehensive identification and characterization of the transcripts, their             expression levels and sub-cellular localizations\n",
      "Comprehensive identification and characterization of the binding sites of             polymerase II\n",
      "Comprehensive identification and characterization of the binding sites of             polymerase II\n",
      "Subsurface mine microbial mat metagenome\n",
      "Oryza sativa Japonica group genome sequencing project by QTL Genomics Research Center, Japan\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "client = OClient()\n",
    "res = client.get('/sra/studies')\n",
    "print(\"total hits: \", res.total)\n",
    "print(\"Titles.............\")\n",
    "for hit in res.hits:\n",
    "    print(hit['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "res2 = client.get('/sra/studies', facets=['center_name', 'study_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "facet: center_name\n",
      "    BioProject : 140823\n",
      "    GEO : 38518\n",
      "    DOE - JOINT GENOME INSTITUTE : 2590\n",
      "    UMIGS : 2557\n",
      "    JGI : 2368\n",
      "    JCVI : 1634\n",
      "    WUGSC : 1402\n",
      "    BI : 976\n",
      "    SC : 969\n",
      "    The Wellcome Trust Sanger Institute : 883\n",
      "facet: study_type\n",
      "    Other : 95040\n",
      "    Whole Genome Sequencing : 66450\n",
      "    Metagenomics : 31662\n",
      "    Transcriptome Analysis : 31087\n",
      "    Population Genomics : 807\n",
      "    Epigenetics : 624\n",
      "    Cancer Genomics : 290\n",
      "    Exome Sequencing : 279\n",
      "    Pooled Clone Sequencing : 31\n",
      "    Synthetic Genomics : 10\n"
     ]
    }
   ],
   "source": [
    "facets = res2.facets\n",
    "for facet_name in res2.facet_names:\n",
    "    print(f'facet: {facet_name}')\n",
    "    facet = res2.facets[facet_name]\n",
    "    for item in facet['buckets']:\n",
    "        print(f\"    {item['key']} : {item['doc_count']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>doc_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BioProject</td>\n",
       "      <td>140823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GEO</td>\n",
       "      <td>38518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOE - JOINT GENOME INSTITUTE</td>\n",
       "      <td>2590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMIGS</td>\n",
       "      <td>2557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JGI</td>\n",
       "      <td>2368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>JCVI</td>\n",
       "      <td>1634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WUGSC</td>\n",
       "      <td>1402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BI</td>\n",
       "      <td>976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SC</td>\n",
       "      <td>969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>The Wellcome Trust Sanger Institute</td>\n",
       "      <td>883</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   key  doc_count\n",
       "0                           BioProject     140823\n",
       "1                                  GEO      38518\n",
       "2         DOE - JOINT GENOME INSTITUTE       2590\n",
       "3                                UMIGS       2557\n",
       "4                                  JGI       2368\n",
       "5                                 JCVI       1634\n",
       "6                                WUGSC       1402\n",
       "7                                   BI        976\n",
       "8                                   SC        969\n",
       "9  The Wellcome Trust Sanger Institute        883"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict(res2.facets['center_name']['buckets'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pubmed_ids</th>\n",
       "      <th>attributes</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>study_type</th>\n",
       "      <th>alias</th>\n",
       "      <th>identifiers</th>\n",
       "      <th>abstract</th>\n",
       "      <th>accession</th>\n",
       "      <th>title</th>\n",
       "      <th>center_name</th>\n",
       "      <th>...</th>\n",
       "      <th>published</th>\n",
       "      <th>received</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>experiment_count</th>\n",
       "      <th>run_count</th>\n",
       "      <th>total_bases</th>\n",
       "      <th>total_spots</th>\n",
       "      <th>mean_bases_per_run</th>\n",
       "      <th>taxon_ids</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[20398357, 25329997]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA38027</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>[{'id': 'PRJDA38027', 'namespace': 'BioProject...</td>\n",
       "      <td>&lt;b&gt;&lt;i&gt;Bacillus subtilis&lt;/i&gt; subsp. &lt;i&gt;natto&lt;/i...</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195 genome ...</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>...</td>\n",
       "      <td>2015-07-31T15:20:44</td>\n",
       "      <td>2009-06-20T02:48:02</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>730668528</td>\n",
       "      <td>10148174</td>\n",
       "      <td>7.306685e+08</td>\n",
       "      <td>[645657]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[20398357]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA39275</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>[{'id': 'PRJDA39275', 'namespace': 'BioProject...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>Model organism for prokaryotic cell differenti...</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-03-24T03:11:55</td>\n",
       "      <td>2009-08-04T07:37:05</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>598805064</td>\n",
       "      <td>8316737</td>\n",
       "      <td>5.988051e+08</td>\n",
       "      <td>[224308]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:53:29</td>\n",
       "      <td>2009-08-06T07:54:04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7461074340</td>\n",
       "      <td>207252065</td>\n",
       "      <td>8.290083e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:54:44</td>\n",
       "      <td>2009-08-06T07:55:04</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1976458608</td>\n",
       "      <td>54901628</td>\n",
       "      <td>6.588195e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:55:26</td>\n",
       "      <td>2009-08-06T07:55:24</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1668749004</td>\n",
       "      <td>46354139</td>\n",
       "      <td>5.562497e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000006</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000006</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:56:11</td>\n",
       "      <td>2009-08-06T07:55:39</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1696349916</td>\n",
       "      <td>47120831</td>\n",
       "      <td>5.654500e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000007</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000007</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:57:44</td>\n",
       "      <td>2009-08-06T07:55:32</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1702328760</td>\n",
       "      <td>47286910</td>\n",
       "      <td>8.511644e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[20400770]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA34559</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DRP000008</td>\n",
       "      <td>[{'id': 'PRJDA34559', 'namespace': 'BioProject...</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>DRP000008</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>...</td>\n",
       "      <td>2010-10-14T04:58:50</td>\n",
       "      <td>2009-08-06T07:55:45</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1643442516</td>\n",
       "      <td>45651181</td>\n",
       "      <td>8.217213e+08</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[22303444]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJNA39577</td>\n",
       "      <td>Metagenomics</td>\n",
       "      <td>DRP000009</td>\n",
       "      <td>[{'id': 'PRJNA39577', 'namespace': 'BioProject...</td>\n",
       "      <td>This is a metagenomic project to figure out th...</td>\n",
       "      <td>DRP000009</td>\n",
       "      <td>Subsurface mine microbial mat metagenome</td>\n",
       "      <td>JAMSTEC</td>\n",
       "      <td>...</td>\n",
       "      <td>2011-03-30T04:00:00</td>\n",
       "      <td>2009-08-21T03:01:03</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>259621154</td>\n",
       "      <td>1016012</td>\n",
       "      <td>8.654038e+07</td>\n",
       "      <td>[527640]</td>\n",
       "      <td>This is a metagenomic project to figure out th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[20423466]</td>\n",
       "      <td>[]</td>\n",
       "      <td>PRJDA39809</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>DRP000010</td>\n",
       "      <td>[{'id': 'PRJDA39809', 'namespace': 'BioProject...</td>\n",
       "      <td>Oryza sativa has important syntenic relationsh...</td>\n",
       "      <td>DRP000010</td>\n",
       "      <td>Oryza sativa Japonica group genome sequencing ...</td>\n",
       "      <td>NIAS</td>\n",
       "      <td>...</td>\n",
       "      <td>2015-07-31T15:20:47</td>\n",
       "      <td>2009-08-21T06:09:02</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8941218259</td>\n",
       "      <td>272124169</td>\n",
       "      <td>9.934687e+08</td>\n",
       "      <td>[39947]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             pubmed_ids attributes  BioProject               study_type  \\\n",
       "0  [20398357, 25329997]         []  PRJDA38027  Whole Genome Sequencing   \n",
       "1            [20398357]         []  PRJDA39275  Whole Genome Sequencing   \n",
       "2            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "3            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "4            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "5            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "6            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "7            [20400770]         []  PRJDA34559   Transcriptome Analysis   \n",
       "8            [22303444]         []  PRJNA39577             Metagenomics   \n",
       "9            [20423466]         []  PRJDA39809  Whole Genome Sequencing   \n",
       "\n",
       "       alias                                        identifiers  \\\n",
       "0  DRP000001  [{'id': 'PRJDA38027', 'namespace': 'BioProject...   \n",
       "1  DRP000002  [{'id': 'PRJDA39275', 'namespace': 'BioProject...   \n",
       "2  DRP000003  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "3  DRP000004  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "4  DRP000005  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "5  DRP000006  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "6  DRP000007  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "7  DRP000008  [{'id': 'PRJDA34559', 'namespace': 'BioProject...   \n",
       "8  DRP000009  [{'id': 'PRJNA39577', 'namespace': 'BioProject...   \n",
       "9  DRP000010  [{'id': 'PRJDA39809', 'namespace': 'BioProject...   \n",
       "\n",
       "                                            abstract  accession  \\\n",
       "0  <b><i>Bacillus subtilis</i> subsp. <i>natto</i...  DRP000001   \n",
       "1                                                NaN  DRP000002   \n",
       "2  Comprehensive identification and characterizat...  DRP000003   \n",
       "3  Comprehensive identification and characterizat...  DRP000004   \n",
       "4  Comprehensive identification and characterizat...  DRP000005   \n",
       "5  Comprehensive identification and characterizat...  DRP000006   \n",
       "6  Comprehensive identification and characterizat...  DRP000007   \n",
       "7  Comprehensive identification and characterizat...  DRP000008   \n",
       "8  This is a metagenomic project to figure out th...  DRP000009   \n",
       "9  Oryza sativa has important syntenic relationsh...  DRP000010   \n",
       "\n",
       "                                               title center_name  ...  \\\n",
       "0  Bacillus subtilis subsp. natto BEST195 genome ...        KEIO  ...   \n",
       "1  Model organism for prokaryotic cell differenti...        KEIO  ...   \n",
       "2  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "3  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "4  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "5  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "6  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "7  Comprehensive identification and characterizat...      UT-MGS  ...   \n",
       "8           Subsurface mine microbial mat metagenome     JAMSTEC  ...   \n",
       "9  Oryza sativa Japonica group genome sequencing ...        NIAS  ...   \n",
       "\n",
       "             published             received sample_count experiment_count  \\\n",
       "0  2015-07-31T15:20:44  2009-06-20T02:48:02            1                1   \n",
       "1  2010-03-24T03:11:55  2009-08-04T07:37:05            1                1   \n",
       "2  2010-10-14T04:53:29  2009-08-06T07:54:04            1                1   \n",
       "3  2010-10-14T04:54:44  2009-08-06T07:55:04            1                1   \n",
       "4  2010-10-14T04:55:26  2009-08-06T07:55:24            1                1   \n",
       "5  2010-10-14T04:56:11  2009-08-06T07:55:39            1                1   \n",
       "6  2010-10-14T04:57:44  2009-08-06T07:55:32            1                2   \n",
       "7  2010-10-14T04:58:50  2009-08-06T07:55:45            1                2   \n",
       "8  2011-03-30T04:00:00  2009-08-21T03:01:03            1                1   \n",
       "9  2015-07-31T15:20:47  2009-08-21T06:09:02            1                1   \n",
       "\n",
       "  run_count total_bases total_spots mean_bases_per_run taxon_ids  \\\n",
       "0         1   730668528    10148174       7.306685e+08  [645657]   \n",
       "1         1   598805064     8316737       5.988051e+08  [224308]   \n",
       "2         9  7461074340   207252065       8.290083e+08    [9606]   \n",
       "3         3  1976458608    54901628       6.588195e+08    [9606]   \n",
       "4         3  1668749004    46354139       5.562497e+08    [9606]   \n",
       "5         3  1696349916    47120831       5.654500e+08    [9606]   \n",
       "6         2  1702328760    47286910       8.511644e+08    [9606]   \n",
       "7         2  1643442516    45651181       8.217213e+08    [9606]   \n",
       "8         3   259621154     1016012       8.654038e+07  [527640]   \n",
       "9         9  8941218259   272124169       9.934687e+08   [39947]   \n",
       "\n",
       "                                         description  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2  Although recent studies have revealed that the...  \n",
       "3  Although recent studies have revealed that the...  \n",
       "4  Although recent studies have revealed that the...  \n",
       "5  Although recent studies have revealed that the...  \n",
       "6  Although recent studies have revealed that the...  \n",
       "7  Although recent studies have revealed that the...  \n",
       "8  This is a metagenomic project to figure out th...  \n",
       "9                                                NaN  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict(res2.hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
